transports/webtransport-websys/src/stream.rs

```diff
@@ -0,0 +1,237 @@
+                let _ = ready!(self.writer_closed_promise.poll(cx));
+                self.writer_state = StreamState::Closed;
+            }
+            StreamState::Closed => {}
+        }
+
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_reader_read(&mut self, cx: &mut Context) -> Poll<io::Result<Option<Uint8Array>>> {
+        let val = ready!(self
+            .reader_read_promise
+            .maybe_init_and_poll(cx, || self.reader.read()))
+        .map_err(to_io_error)?;
+
+        let val = parse_reader_response(&val)
+            .map_err(to_io_error)?
+            .map(Uint8Array::from);
+
+        Poll::Ready(Ok(val))
+    }
+}
+
+impl Drop for Stream {
+    fn drop(&mut self) {
+        // On drop we abort any writes and cancel any ongoing reads
+        detach_promise(self.writer.abort());
+        detach_promise(self.reader.cancel());
+    }
+}
+
+impl AsyncRead for Stream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        let this = self.get_mut();
+
+        // If we have leftovers from a previous read, then use them.
+        // Otherwise read new data.
+        let data = match this.read_leftovers.take() {
+            Some(data) => data,
+            None => {
+                match ready!(this.poll_reader_read(cx))? {
+                    Some(data) => data,
+                    // EOF
+                    None => return Poll::Ready(Ok(0)),
+                }
+            }
+        };
+
+        if data.byte_length() == 0 {
+            return Poll::Ready(Ok(0));
+        }
+
+        let out_len = data.byte_length().min(buf.len() as u32);
+        data.slice(0, out_len).copy_to(&mut buf[..out_len as usize]);
+
+        let leftovers = data.slice(out_len, data.byte_length());
+
+        if leftovers.byte_length() > 0 {
+            this.read_leftovers = Some(leftovers);
+        }
+
+        Poll::Ready(Ok(out_len as usize))
+    }
+}
+
+impl AsyncWrite for Stream {
+    fn poll_write(self: Pin<&mut Self>, cx: &mut Context, buf: &[u8]) -> Poll<io::Result<usize>> {
+        let this = self.get_mut();
+
+        ready!(this.poll_writer_ready(cx))?;
+
+        let len = buf.len() as u32;
+        let data = Uint8Array::new_with_length(len);
+        data.copy_from(buf);
+
+        detach_promise(this.writer.write_with_chunk(&data));
+
+        Poll::Ready(Ok(len as usize))
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context) -> Poll<io::Result<()>> {
+        let this = self.get_mut();
+
+        if this.writer_state == StreamState::Open {
+            // Writer has queue size of 1, so as soon it is ready, this means the
+            // messages were flushed.
+            this.poll_writer_ready(cx)
+        } else if cfg!(debug_assertions) {
+            panic!("libp2p_webtransport_websys::Stream: poll_flush called after poll_close");
+        } else {
+            Poll::Ready(Ok(()))
+        }
```

Comment:
```suggestion
        } else {
            debug_assert!(
                false,
                "libp2p_webtransport_websys::Stream: poll_flush called after poll_close"
            );
            Poll::Ready(Ok(()))
        }
```

I would find this easier to read. What do you think? No strong preference.